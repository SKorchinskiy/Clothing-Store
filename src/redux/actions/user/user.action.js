import { USER_ACTION_TYPES } from "./user.type";
import { createAction } from "../create-action.helper";
import {
  signInUserWithGoogle,
  signInUserByEmail,
  signUpUserByEmail,
  signOutUser,
} from "../../../configs/firebase.config";

export function setCurrentUser(user) {
  return createAction(USER_ACTION_TYPES.SET_USER, user);
}

function fetchCurrentUserStart() {
  return createAction(USER_ACTION_TYPES.FETCH_CURRENT_USER_START);
}

function fetchCurrentUserSuccess(currentUser) {
  return createAction(
    USER_ACTION_TYPES.FETCH_CURRENT_USER_SUCCESS,
    currentUser
  );
}

function fetchCurrentUserFailed(error) {
  return createAction(USER_ACTION_TYPES.FETCH_CURRENT_USER_FAILED, error);
}

export function startEmailSignUp({ name, email, password }) {
  return async (dispatch) => {
    dispatch(fetchCurrentUserStart());
    try {
      const currentUser = await signUpUserByEmail({ name, email, password });
      dispatch(fetchCurrentUserSuccess(currentUser));
    } catch (error) {
      dispatch(fetchCurrentUserFailed(error));
    }
  };
}

export function startGoogleSignIn() {
  return async (dispatch) => {
    dispatch(fetchCurrentUserStart());
    try {
      const currentUser = await signInUserWithGoogle();
      dispatch(fetchCurrentUserSuccess(currentUser));
    } catch (error) {
      dispatch(fetchCurrentUserFailed(error));
    }
  };
}

export function startEmailSignIn(email, password) {
  return async (dispatch) => {
    dispatch(fetchCurrentUserStart);
    try {
      const currentUser = await signInUserByEmail(email, password);
      dispatch(fetchCurrentUserSuccess(currentUser));
    } catch (error) {
      dispatch(fetchCurrentUserFailed(error));
    }
  };
}

export function signOutCurrentUserAsync() {
  return async (dispatch) => {
    dispatch(fetchCurrentUserStart());
    try {
      await signOutUser();
      dispatch(setCurrentUser(null));
    } catch (error) {
      dispatch(fetchCurrentUserFailed(error));
    }
  };
}
