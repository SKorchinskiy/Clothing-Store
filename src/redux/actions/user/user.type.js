export const USER_ACTION_TYPES = {
  SET_USER: "SET_USER",
};
