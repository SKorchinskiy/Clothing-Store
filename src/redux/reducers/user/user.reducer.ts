import { AnyAction } from "redux";

import {
  setCurrentUser,
  startEmailSignIn,
  startEmailSignUp,
  startGoogleSignIn,
  startSignOut,
  signInFailed,
  signInSuccess,
  signOutFailed,
  signOutSuccess,
  signUpFailed,
  signUpSuccess,
} from "../../actions/user/user.action";

export type User = {
  displayName: string;
  email: string;
};

export type UserState = {
  readonly currentUser: User | null;
  readonly isLoading: boolean;
  readonly error: Error | null;
};

const INITIAL_USER_STATE: UserState = {
  currentUser: null,
  isLoading: false,
  error: null,
};

export function userReducer(state = INITIAL_USER_STATE, action: AnyAction) {
  if (setCurrentUser.match(action)) {
    return {
      ...state,
      isLoading: false,
      currentUser: action.payload,
    };
  }

  if (
    startEmailSignIn.match(action) ||
    startGoogleSignIn.match(action) ||
    startEmailSignUp.match(action) ||
    startSignOut.match(action)
  ) {
    return {
      ...state,
      isLoading: true,
    };
  }

  if (
    signInSuccess.match(action) ||
    signUpSuccess.match(action) ||
    signOutSuccess.match(action)
  ) {
    return {
      ...state,
      isLoading: false,
      currentUser: action.payload,
    };
  }

  if (
    signInFailed.match(action) ||
    signUpFailed.match(action) ||
    signOutFailed.match(action)
  ) {
    return {
      ...state,
      isLoading: false,
      error: action.payload,
    };
  }

  return state;
}
