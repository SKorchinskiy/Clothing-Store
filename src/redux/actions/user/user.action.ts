import { USER_ACTION_TYPES } from "./user.type";
import {
  Action,
  ActionWithPayload,
  createAction,
  withMatcher,
} from "../create-action.helper";
import { User } from "../../reducers/user/user.reducer";

export type SignInUser = {
  email: string;
  password: string;
};

export type SignUpUser = SignInUser & { name: string };

export type SetCurrentUser = ActionWithPayload<
  USER_ACTION_TYPES.SET_CURRENT_USER,
  User
>;

export const setCurrentUser = withMatcher(
  (user: User): SetCurrentUser =>
    createAction(USER_ACTION_TYPES.SET_CURRENT_USER, user)
);

export type CheckCurrentUserSession =
  Action<USER_ACTION_TYPES.CHECK_USER_SESSION>;

export const checkCurrentUserSession = withMatcher(
  (): CheckCurrentUserSession =>
    createAction(USER_ACTION_TYPES.CHECK_USER_SESSION)
);

export type StartGoogleSignIn = Action<USER_ACTION_TYPES.GOOGLE_SIGN_IN_START>;

export const startGoogleSignIn = withMatcher(
  (): StartGoogleSignIn => createAction(USER_ACTION_TYPES.GOOGLE_SIGN_IN_START)
);

export type StartEmailSignIn = ActionWithPayload<
  USER_ACTION_TYPES.EMAIL_SIGN_IN_START,
  SignInUser
>;

export const startEmailSignIn = withMatcher(
  (email: string, password: string): StartEmailSignIn =>
    createAction(USER_ACTION_TYPES.EMAIL_SIGN_IN_START, {
      email,
      password,
    })
);

export type StartEmailSignUp = ActionWithPayload<
  USER_ACTION_TYPES.EMAIL_SIGN_UP_START,
  SignUpUser
>;

export const startEmailSignUp = withMatcher(
  (data: SignUpUser): StartEmailSignUp =>
    createAction(USER_ACTION_TYPES.EMAIL_SIGN_UP_START, data)
);

export type StartSignOut = Action<USER_ACTION_TYPES.SIGN_OUT_START>;

export const startSignOut = withMatcher(
  (): StartSignOut => createAction(USER_ACTION_TYPES.SIGN_OUT_START)
);

export type SignInSuccess = ActionWithPayload<
  USER_ACTION_TYPES.SIGN_IN_SUCCESS,
  User
>;

export const signInSuccess = withMatcher(
  (currentUser: User): SignInSuccess =>
    createAction(USER_ACTION_TYPES.SIGN_IN_SUCCESS, currentUser)
);

export type SignUpSuccess = ActionWithPayload<
  USER_ACTION_TYPES.SIGN_UP_SUCCESS,
  User
>;

export const signUpSuccess = withMatcher(
  (currentUser: User): SignUpSuccess =>
    createAction(USER_ACTION_TYPES.SIGN_UP_SUCCESS, currentUser)
);

export type SignOutSuccess = ActionWithPayload<
  USER_ACTION_TYPES.SIGN_OUT_SUCCESS,
  null
>;

export const signOutSuccess = withMatcher(
  (): SignOutSuccess => createAction(USER_ACTION_TYPES.SIGN_OUT_SUCCESS, null)
);

export type SignInFailed = ActionWithPayload<
  USER_ACTION_TYPES.SIGN_IN_FAILED,
  Error
>;

export const signInFailed = withMatcher(
  (error: Error): SignInFailed =>
    createAction(USER_ACTION_TYPES.SIGN_IN_FAILED, error)
);

export type SignUpFailed = ActionWithPayload<
  USER_ACTION_TYPES.SIGN_UP_FAILED,
  Error
>;

export const signUpFailed = withMatcher(
  (error: Error): SignUpFailed =>
    createAction(USER_ACTION_TYPES.SIGN_UP_FAILED, error)
);

export type SignOutFailed = ActionWithPayload<
  USER_ACTION_TYPES.SIGN_OUT_FAILED,
  Error
>;

export const signOutFailed = withMatcher(
  (error: Error): SignOutFailed =>
    createAction(USER_ACTION_TYPES.SIGN_OUT_FAILED, error)
);
