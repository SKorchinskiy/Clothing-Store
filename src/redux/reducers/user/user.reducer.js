import { USER_ACTION_TYPES } from "../../actions/user/user.type";

const INITIAL_USER_STATE = {
  currentUser: null,
};

export function userReducer(state = INITIAL_USER_STATE, action) {
  const { type, payload } = action;
  switch (type) {
    case USER_ACTION_TYPES.SET_USER: {
      return {
        ...state,
        currentUser: payload,
      };
    }
    default: {
      return state;
    }
  }
}
