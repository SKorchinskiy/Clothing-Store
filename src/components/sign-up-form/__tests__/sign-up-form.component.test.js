import "@testing-library/jest-dom";
import { act, fireEvent, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import SignInForm from "../sign-up-form.component";

import { userStub } from "./stubs/sign-up-user.stub";

import renderWithProviders from "../../../utils/test.utils";
import { startEmailSignUp } from "../../../redux/actions/user/user.action";

const mockedUseDispatch = jest.fn();

jest.mock("react-redux", () => {
  const actualModule = jest.requireActual("react-redux");
  return {
    ...actualModule,
    useDispatch: () => {
      const actualUseDispatch = actualModule.useDispatch();
      return (action) => {
        actualUseDispatch(action);
        mockedUseDispatch(action);
      };
    },
  };
});

describe("<SignInForm /> component", () => {
  it("should render initial UI properly", () => {
    const { asFragment } = renderWithProviders(<SignInForm />);

    const nameElement = screen.getByLabelText("name");
    expect(nameElement).toBeInTheDocument();

    const emailElement = screen.getByLabelText("email");
    expect(emailElement).toBeInTheDocument();

    const passwordElement = screen.getByLabelText("password");
    expect(passwordElement).toBeInTheDocument();

    const confirmPasswordElement = screen.getByLabelText("confirm password");
    expect(confirmPasswordElement).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should dispatch startEmailSignUp when button is clicked", async () => {
    renderWithProviders(<SignInForm />);

    const nameElement = screen.getByLabelText("name");

    await act(async () => await userEvent.click(nameElement));
    await act(async () => await userEvent.paste(userStub.name));
    await waitFor(() => expect(nameElement.value).toBe(userStub.name));

    const emailElement = screen.getByLabelText("email");

    await act(async () => await userEvent.click(emailElement));
    await act(async () => await userEvent.paste(userStub.email));
    await waitFor(() => expect(emailElement.value).toBe(userStub.email));

    const passwordElement = screen.getByLabelText("password");

    await act(async () => await userEvent.click(passwordElement));
    await act(async () => await userEvent.paste(userStub.password));
    await waitFor(() => expect(passwordElement.value).toBe(userStub.password));

    const confirmPasswordElement = screen.getByLabelText("confirm password");
    await act(async () => await userEvent.click(confirmPasswordElement));
    await act(async () => await userEvent.paste(userStub.confirmPassword));
    await waitFor(() =>
      expect(confirmPasswordElement.value).toBe(userStub.confirmPassword)
    );

    const signUpButton = screen.getByText(/sign up/i, { selector: "button" });
    await act(async () => await userEvent.click(signUpButton));

    await waitFor(() =>
      expect(mockedUseDispatch).toHaveBeenCalledWith(
        startEmailSignUp({
          name: userStub.name,
          email: userStub.email,
          password: userStub.password,
        })
      )
    );

    expect(mockedUseDispatch).toHaveBeenCalledTimes(1);
  });
});
