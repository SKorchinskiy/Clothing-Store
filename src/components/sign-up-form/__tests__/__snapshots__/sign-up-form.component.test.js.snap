// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignInForm /> component should render initial UI properly 1`] = `
<DocumentFragment>
  <div
    class="sc-bcPKhP edaQst"
  >
    <div
      class="sc-dkjKgF fBdIER"
    >
      <h3>
        I don't have an account
      </h3>
      <p>
        Sign Up with your email and password
      </p>
    </div>
    <form
      class="sc-grXZZQ kywDov"
    >
      <div
        class="sc-ftLKQv fhLrxm"
      >
        <label
          for="signup-name"
        >
          name
        </label>
        <input
          class="sc-iBAaJG fZYlVh"
          id="signup-name"
          name="name"
          placeholder="enter a name"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="sc-ftLKQv fhLrxm"
      >
        <label
          for="signup-email"
        >
          email
        </label>
        <input
          class="sc-iBAaJG fZYlVh"
          id="signup-email"
          name="email"
          placeholder="enter an email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="sc-ftLKQv fhLrxm"
      >
        <label
          for="signup-password"
        >
          password
        </label>
        <input
          class="sc-iBAaJG fZYlVh"
          id="signup-password"
          name="password"
          placeholder="enter a password"
          required=""
          type="password"
          value=""
        />
      </div>
      <div
        class="sc-ftLKQv fhLrxm"
      >
        <label
          for="signup-confirmPassword"
        >
          confirm password
        </label>
        <input
          class="sc-iBAaJG fZYlVh"
          id="signup-confirmPassword"
          name="confirmPassword"
          placeholder="confirm a password"
          required=""
          type="password"
          value=""
        />
      </div>
      <button
        class="sc-hLclGa sc-eDnVMP ibGimQ jZOKtP"
        type="submit"
      >
        Sign Up
      </button>
    </form>
  </div>
</DocumentFragment>
`;
